### ISSUE TITLE
does nncf support post training quantization and jupyter notebook? 

### ISSUE THREAD

**Issue by liwei46** [openvinotoolkit/nncf_pytorch#78]
Hello,
Thanks for this great project. But
1) does nncf support post training quantization? I did not find related information/example yet.
2) it seems nncf does not  jupyter notebook?  There will be error if import modules in notebook. 

Thanks.

**Comment by vshampor** [comment]
Greetings, @liwei46!

1. 

The NNCF is based upon PyTorch, which itself is a *training* framework. Now, you _could_ possibly use it just to import models and convert them to ONNX (if you have no GPU, for instance), but this does not seem to be the major use case for PyTorch. 

Anyway, it _is_ possible to do post-training quantization with NNCF, in a sense. Just specify the `"range"` and the `"batchnorm_adaptation"` initializer sections in the NNCF config as described in the docs: [one](https://github.com/openvinotoolkit/nncf_pytorch/blob/develop/docs/ConfigFile.md#nncf-configuration-file-description), [two](https://github.com/openvinotoolkit/nncf_pytorch/blob/develop/docs/compression_algorithms/Quantization.md). Note that you should supply a dataset for the initialization to happen (you can use a [helper function](https://github.com/openvinotoolkit/nncf_pytorch/blob/caca391e81d0aa62f013d3f7425755d6679b0a03/nncf/initialization.py#L219) for that), since initializing activation quantizers entails running the model inference one or multiple times. Once you do that, during the `create_compressed_model` call both initializations will happen while only using forward passes of the model (no backward), and the model that is output from that `create_compressed_model` will then contain all quantizers configured as best as currently possible with NNCF without running training. You can try exporting the model to ONNX at that point right away, or running evaluation scripts directly in PyTorch to see if the resulting no-training compressed model quality is suitable for your purposes.

Still, for best results, I would recommend to actually train the NNCF-wrapped model obtained after the `create_compressed_model` call so that the quantizer parameters and the model weights can adapt together to the low-precision execution scenario. This is what NNCF specializes in.

If you uncompromisingly require post-training quantization for OpenVINO-powered inference, then you can use a tool that is specialized exactly in that - the Post Training Optimization Toolkit, distributed as part of the OpenVINO packages. It has a larger scale of algorithms that specifically improve post-training quantization performance and quality. See [POT documentation](https://docs.openvinotoolkit.org/latest/pot_README.html) and general OpenVINO docs for instructions on how to use it.

2. 

Could you please give more details on the errors during Jupyter import and post logs/versions of NNCF and Jupyter you used? We will be happy to look into that and enable Jupyter notebook imports if for any reason it is broken right now.

**Comment by liwei46** [comment]
Thanks for the quick and helpful reply. 
Below is the error message when I type:
 `from nncf.dynamic_graph.version_agnostic_op_names import VersionAgnosticNames` in jupyter notebook.
```
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-1-76d95093b247> in <module>
----> 1 from nncf.dynamic_graph.version_agnostic_op_names import VersionAgnosticNames

~/anaconda3/envs/nncf/lib/python3.7/site-packages/nncf-1.3.2-py3.7.egg/nncf/__init__.py in <module>
     15 
     16 # Required for correct COMPRESSION_ALGORITHMS registry functioning
---> 17 from .binarization import algo as binarization_algo
     18 from .quantization import algo as quantization_algo
     19 from .sparsity.const import algo as const_sparsity_algo

~/anaconda3/envs/nncf/lib/python3.7/site-packages/nncf-1.3.2-py3.7.egg/nncf/binarization/algo.py in <module>
     18 from torch import nn
     19 
---> 20 from nncf.algo_selector import COMPRESSION_ALGORITHMS
     21 from nncf.binarization.layers import BINARIZATION_MODULES, BinarizationMode, WeightBinarizer, ActivationBinarizer, \
     22     ActivationBinarizationScaleThreshold, BaseBinarizer

~/anaconda3/envs/nncf/lib/python3.7/site-packages/nncf-1.3.2-py3.7.egg/nncf/algo_selector.py in <module>
     12 """
     13 
---> 14 from .compression_method_api import CompressionAlgorithmBuilder, CompressionAlgorithmController
     15 from .registry import Registry
     16 

~/anaconda3/envs/nncf/lib/python3.7/site-packages/nncf-1.3.2-py3.7.egg/nncf/compression_method_api.py in <module>
     27 from nncf.config import NNCFConfig
     28 from nncf.dynamic_graph.graph_builder import create_mock_tensor
---> 29 from nncf.initialization import DataLoaderBNAdaptationRunner
     30 from nncf.nncf_logger import logger as nncf_logger
     31 from nncf.nncf_network import NNCFNetwork

~/anaconda3/envs/nncf/lib/python3.7/site-packages/nncf-1.3.2-py3.7.egg/nncf/initialization.py in <module>
      9 
     10 from nncf.nncf_logger import logger as nncf_logger
---> 11 from nncf.quantization.init_range import MinMaxInitializer, ThreeSigmaInitializer, MeanMinMaxInitializer
     12 from nncf.quantization.init_range import PercentileInitializer
     13 from nncf.structures import QuantizationPrecisionInitArgs, QuantizationRangeInitArgs, BNAdaptationInitArgs

~/anaconda3/envs/nncf/lib/python3.7/site-packages/nncf-1.3.2-py3.7.egg/nncf/quantization/__init__.py in <module>
     12 """
     13 # Required for correct QUANTIZATION_MODULES registry functioning
---> 14 from . import layers

~/anaconda3/envs/nncf/lib/python3.7/site-packages/nncf-1.3.2-py3.7.egg/nncf/quantization/layers.py in <module>
     23 from nncf.functions import clamp
     24 from nncf.nncf_logger import logger as nncf_logger
---> 25 from .quantize_functions import symmetric_quantize, asymmetric_quantize, ExportQuantizeToFakeQuantize, \
     26     get_scale_zp_from_input_low_input_high, ExportQuantizeToONNXQuantDequant, TuneRange
     27 from ..layer_utils import COMPRESSION_MODULES

~/anaconda3/envs/nncf/lib/python3.7/site-packages/nncf-1.3.2-py3.7.egg/nncf/quantization/quantize_functions.py in <module>
     14 import warnings
     15 
---> 16 from .extensions import QuantizedFunctionsCPU, QuantizedFunctionsCUDA
     17 from ..dynamic_graph.patch_pytorch import register_operator
     18 from ..functions import STRound, clamp

~/anaconda3/envs/nncf/lib/python3.7/site-packages/nncf-1.3.2-py3.7.egg/nncf/quantization/extensions.py in <module>
     39 QuantizedFunctionsCPU = load(
     40     'quantized_functions_cpu', CPU_EXT_SRC_LIST, extra_include_paths=EXT_INCLUDE_DIRS,
---> 41     verbose=False
     42 )
     43 

~/anaconda3/envs/nncf/lib/python3.7/site-packages/torch-1.5.0-py3.7-linux-x86_64.egg/torch/utils/cpp_extension.py in load(name, sources, extra_cflags, extra_cuda_cflags, extra_ldflags, extra_include_paths, build_directory, verbose, with_cuda, is_python_module)
    896         verbose,
    897         with_cuda,
--> 898         is_python_module)
    899 
    900 

~/anaconda3/envs/nncf/lib/python3.7/site-packages/torch-1.5.0-py3.7-linux-x86_64.egg/torch/utils/cpp_extension.py in _jit_compile(name, sources, extra_cflags, extra_cuda_cflags, extra_ldflags, extra_include_paths, build_directory, verbose, with_cuda, is_python_module)
   1063         build_arguments=[extra_cflags, extra_cuda_cflags, extra_ldflags, extra_include_paths],
   1064         build_directory=build_directory,
-> 1065         with_cuda=with_cuda
   1066     )
   1067     if version > 0:

~/anaconda3/envs/nncf/lib/python3.7/site-packages/torch-1.5.0-py3.7-linux-x86_64.egg/torch/utils/_cpp_extension_versioner.py in bump_version_if_changed(self, name, source_files, build_arguments, build_directory, with_cuda)
     42                                 with_cuda):
     43         hash_value = 0
---> 44         hash_value = hash_source_files(hash_value, source_files)
     45         hash_value = hash_build_arguments(hash_value, build_arguments)
     46         hash_value = update_hash(hash_value, build_directory)

~/anaconda3/envs/nncf/lib/python3.7/site-packages/torch-1.5.0-py3.7-linux-x86_64.egg/torch/utils/_cpp_extension_versioner.py in hash_source_files(hash_value, source_files)
     14 def hash_source_files(hash_value, source_files):
     15     for filename in source_files:
---> 16         with open(filename) as file:
     17             hash_value = update_hash(hash_value, file.read())
     18     return hash_value

FileNotFoundError: [Errno 2] No such file or directory: '/home/wei_li/anaconda3/envs/nncf/lib/python3.7/site-packages/nncf-1.3.2-py3.7.egg/nncf/extensions/src/quantization/cpu/functions_cpu.cpp'
```

**Comment by vshampor** [comment]
@liwei46 try uninstalling NNCF and reinstalling the latest version from develop (or the v1.4.1 release version at least). There have been some packaging problems with the v1.3.2 of the exact kind that you are describing, that should be fixed starting from v1.4.

**Comment by liwei46** [comment]
thanks.
